File: backend/app/routers/annotations.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from pathlib import Path

from ..database import get_db
from ..models import ImageRecord, Annotation
from ..schemas import (
    AnnotationCreate,
    AnnotationUpdate,
    AnnotationResponse,
    ImageAnnotations,
)
from ..config import settings

router = APIRouter(prefix="/api/annotations", tags=["annotations"])


def get_or_create_image(db: Session, filename: str) -> ImageRecord:
    """Get existing image record or create new one."""
    image = db.query(ImageRecord).filter(ImageRecord.filename == filename).first()
    if image:
        return image

    # Check if file exists
    filepath = Path(settings.images_dir) / filename
    if not filepath.exists():
        raise HTTPException(status_code=404, detail="Image file not found")

    # Get dimensions
    from PIL import Image as PILImage
    with PILImage.open(filepath) as img:
        width, height = img.size

    image = ImageRecord(filename=filename, width=width, height=height)
    db.add(image)
    db.commit()
    db.refresh(image)
    return image


@router.get("/{filename}", response_model=ImageAnnotations)
async def get_annotations(filename: str, db: Session = Depends(get_db)):
    """Get all annotations for an image."""
    image = db.query(ImageRecord).filter(ImageRecord.filename == filename).first()

    if not image:
        # Check if file exists, create record if so
        filepath = Path(settings.images_dir) / filename
        if not filepath.exists():
            raise HTTPException(status_code=404, detail="Image not found")
        image = get_or_create_image(db, filename)

    annotations = [
        AnnotationResponse(
            id=ann.id,
            x=ann.x,
            y=ann.y,
            width=ann.width,
            height=ann.height,
            transcription=ann.transcription,
            class_id=ann.class_id,
            created_at=ann.created_at,
            updated_at=ann.updated_at,
        )
        for ann in image.annotations
    ]

    return ImageAnnotations(
        filename=image.filename,
        width=image.width,
        height=image.height,
        annotations=annotations,
    )


@router.post("/{filename}", response_model=AnnotationResponse)
async def create_annotation(
    filename: str, annotation: AnnotationCreate, db: Session = Depends(get_db)
):
    """Create a new annotation."""
    image = get_or_create_image(db, filename)

    db_annotation = Annotation(
        image_id=image.id,
        x=annotation.x,
        y=annotation.y,
        width=annotation.width,
        height=annotation.height,
        transcription=annotation.transcription,
        class_id=annotation.class_id,
    )
    db.add(db_annotation)
    db.commit()
    db.refresh(db_annotation)

    return AnnotationResponse(
        id=db_annotation.id,
        x=db_annotation.x,
        y=db_annotation.y,
        width=db_annotation.width,
        height=db_annotation.height,
        transcription=db_annotation.transcription,
        class_id=db_annotation.class_id,
        created_at=db_annotation.created_at,
        updated_at=db_annotation.updated_at,
    )


@router.put("/{filename}/{annotation_id}", response_model=AnnotationResponse)
async def update_annotation(
    filename: str,
    annotation_id: int,
    annotation: AnnotationUpdate,
    db: Session = Depends(get_db),
):
    """Update an existing annotation."""
    db_annotation = db.query(Annotation).filter(Annotation.id == annotation_id).first()
    if not db_annotation:
        raise HTTPException(status_code=404, detail="Annotation not found")

    update_data = annotation.model_dump(exclude_unset=True)
    for field, value in update_data.items():
        setattr(db_annotation, field, value)

    db.commit()
    db.refresh(db_annotation)

    return AnnotationResponse(
        id=db_annotation.id,
        x=db_annotation.x,
        y=db_annotation.y,
        width=db_annotation.width,
        height=db_annotation.height,
        transcription=db_annotation.transcription,
        class_id=db_annotation.class_id,
        created_at=db_annotation.created_at,
        updated_at=db_annotation.updated_at,
    )


@router.delete("/{filename}/{annotation_id}")
async def delete_annotation(
    filename: str, annotation_id: int, db: Session = Depends(get_db)
):
    """Delete an annotation."""
    db_annotation = db.query(Annotation).filter(Annotation.id == annotation_id).first()
    if not db_annotation:
        raise HTTPException(status_code=404, detail="Annotation not found")

    db.delete(db_annotation)
    db.commit()
    return {"status": "deleted", "id": annotation_id}


@router.post("/{filename}/bulk", response_model=ImageAnnotations)
async def save_all_annotations(
    filename: str, annotations: list[AnnotationCreate], db: Session = Depends(get_db)
):
    """Save all annotations for an image (replaces existing)."""
    image = get_or_create_image(db, filename)

    # Delete existing annotations
    db.query(Annotation).filter(Annotation.image_id == image.id).delete()

    # Create new annotations
    new_annotations = []
    for ann in annotations:
        db_annotation = Annotation(
            image_id=image.id,
            x=ann.x,
            y=ann.y,
            width=ann.width,
            height=ann.height,
            transcription=ann.transcription,
            class_id=ann.class_id,
        )
        db.add(db_annotation)
        new_annotations.append(db_annotation)

    db.commit()

    # Refresh to get IDs
    for ann in new_annotations:
        db.refresh(ann)

    return ImageAnnotations(
        filename=image.filename,
        width=image.width,
        height=image.height,
        annotations=[
            AnnotationResponse(
                id=ann.id,
                x=ann.x,
                y=ann.y,
                width=ann.width,
                height=ann.height,
                transcription=ann.transcription,
                class_id=ann.class_id,
                created_at=ann.created_at,
                updated_at=ann.updated_at,
            )
            for ann in new_annotations
        ],
    )

File: frontend/src/api/annotations.ts
import { fetchJSON } from './client';
import { BoundingBox, ImageAnnotations } from '../types';

interface AnnotationResponse {
  id: number;
  x: number;
  y: number;
  width: number;
  height: number;
  transcription: string;
  class_id: number;
  created_at: string;
  updated_at: string;
}

interface ImageAnnotationsResponse {
  filename: string;
  width: number;
  height: number;
  annotations: AnnotationResponse[];
}

function mapAnnotation(ann: AnnotationResponse): BoundingBox {
  return {
    id: ann.id,
    x: ann.x,
    y: ann.y,
    width: ann.width,
    height: ann.height,
    transcription: ann.transcription,
    classId: ann.class_id,
  };
}

export async function fetchAnnotations(
  filename: string
): Promise<ImageAnnotations> {
  const data = await fetchJSON<ImageAnnotationsResponse>(
    `/annotations/${encodeURIComponent(filename)}`
  );
  return {
    filename: data.filename,
    width: data.width,
    height: data.height,
    annotations: data.annotations.map(mapAnnotation),
  };
}

export async function saveAnnotations(
  filename: string,
  annotations: BoundingBox[]
): Promise<ImageAnnotations> {
  const payload = annotations.map((ann) => ({
    x: ann.x,
    y: ann.y,
    width: ann.width,
    height: ann.height,
    transcription: ann.transcription,
    class_id: ann.classId || 0,
  }));

  const data = await fetchJSON<ImageAnnotationsResponse>(
    `/annotations/${encodeURIComponent(filename)}/bulk`,
    {
      method: 'POST',
      body: JSON.stringify(payload),
    }
  );

  return {
    filename: data.filename,
    width: data.width,
    height: data.height,
    annotations: data.annotations.map(mapAnnotation),
  };
}

export async function deleteAnnotation(
  filename: string,
  annotationId: number
): Promise<void> {
  await fetchJSON(`/annotations/${encodeURIComponent(filename)}/${annotationId}`, {
    method: 'DELETE',
  });
}
